packages/react-reconciler/src/ReactFiberCommitWork.new.js

```diff
@@ -293,7 +295,15 @@ function commitBeforeMutationLifeCycles(
       }
       return;
     }
-    case HostRoot:
+    case HostRoot: {
+      if (supportsMutation) {
+        if (finishedWork.effectTag & ClearContainer) {
+          const root = finishedWork.stateNode;
+          clearContainer(root.containerInfo);
```

Comment:
So I get that the problem is that if you do this in the mutation phase you’ll delete the ones we just added. However if we do it in the snapshot, then these snapshot life cycles and beforeblur observe an already mutated tree and can’t read from the true previous state and it will likely trigger an additional layout pass if they read the layout.

I wonder if we should instead just have the clearContainer traverse top down until it hits the first react managed node we know about.

Alternatively we could special case this and call it between the snapshot and mutation phases. However then that gets tricky if we also want to do this for portals.

Reply:
I guess it would be after the snapshots since the root is last but it wouldn’t work for portals.

Reply:
Yeah, special casing also felt kind of dirty.

As for handling it during snapshot, I mentioned that in my comment above. We *could* do that but it feels fragile.

Reply:
Can we at least just use the Snapshot effect tag for this case?

I want to use the tags as a hint which children to traverse into when we switch to the traversal for effects.

So Snapshot would be the one we look for to enter this path.

Reply:
Ah. That sounds promising. Let me give it a shot.

Reply:
Yes, that will work nicely and avoid the extra effect tag. Good suggestion.